src/iceberg/type.cc

```diff
@@ -0,0 +1,314 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "iceberg/type.h"
+
+#include <format>
+#include <iterator>
+#include <stdexcept>
+
+#include "iceberg/util/formatter.h"
+
+namespace iceberg {
+
+TypeId BooleanType::type_id() const { return TypeId::kBoolean; }
+std::string BooleanType::ToString() const { return "boolean"; }
+bool BooleanType::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kBoolean;
+}
+
+TypeId Int32Type::type_id() const { return TypeId::kInt32; }
+std::string Int32Type::ToString() const { return "int32"; }
+bool Int32Type::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kInt32;
+}
+
+TypeId Int64Type::type_id() const { return TypeId::kInt64; }
+std::string Int64Type::ToString() const { return "int64"; }
+bool Int64Type::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kInt64;
+}
+
+TypeId Float32Type::type_id() const { return TypeId::kFloat32; }
+std::string Float32Type::ToString() const { return "float32"; }
+bool Float32Type::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kFloat32;
+}
+
+TypeId Float64Type::type_id() const { return TypeId::kFloat64; }
+std::string Float64Type::ToString() const { return "float64"; }
+bool Float64Type::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kFloat64;
+}
+
+DecimalType::DecimalType(int32_t precision, int32_t scale)
+    : precision_(precision), scale_(scale) {
+  if (precision < 0 || precision > kMaxPrecision) {
+    throw std::runtime_error(
+        std::format("DecimalType: precision must be in [0, 38], was {}", precision));
+  }
```

Comment:
Now that we've checked the precision, should we also check the scale here?

Reply:
Java actually doesn't validate the scale, and the spec is silent on this part. (Also, Java actually allows negative precision, though I don't think that makes any sense?)

Reply:
From https://github.com/apache/iceberg/issues/7420 I think setting it or not is all possible

Reply:
Blue give some examples which scale can be greater than precision and smaller than 0[1], though I don't think it makes sense, I agree we should keep this as is.

[1] https://github.com/apache/iceberg/issues/7420#issuecomment-1526522715

Reply:
Can we fix this in the spec? The spec itself says nothing about these types which is a little frustrating since I feel like the Java implementation is the actual de-facto spec

Reply:
Oh wow, the issue asking for spec clarification was closed. Alright then, I will leave this as-is